### PR TITLE
fix: don't follow symlinks to execroot

### LINKED ISSUE #920: [Bug]: js_image_layer.mjs: Error: couldn't map o to a path.
### What happened?

During invocation of js_image_layer, I get this eror message:
```
file:///home/jpgariep/.cache/bazel/_bazel_jpgariep/dc2fb24bedc6b4555f41da611ff4d0e4/sandbox/linux-sandbox/1030/execroot/nu_workforce/bazel-out/k8-opt-exec-2B5CBBC6-ST-625e526ca8a8/bin/external/aspect_rules_js/js/private/js_image_layer_builder.sh.runfiles/aspect_rules_js/js/private/js_image_layer.mjs:9578
    throw new Error(`couldn't map ${value} to a path. please file a bug at https://github.com/aspect-build/rules_js/issues/new/choose`);

Error: couldn't map o to a path. please file a bug at https://github.com/aspect-build/rules_js/issues/new/choose
    at findKeyByValue (file:///home/jpgariep/.cache/bazel/_bazel_jpgariep/dc2fb24bedc6b4555f41da611ff4d0e4/sandbox/linux-sandbox/1030/execroot/nu_workforce/bazel-out/k8-opt-exec-2B5CBBC6-ST-625e526ca8a8/bin/external/aspect_rules_js/js/private/js_image_layer_builder.sh.runfiles/aspect_rules_js/js/private/js_image_layer.mjs:9578:11)
    at build (file:///home/jpgariep/.cache/bazel/_bazel_jpgariep/dc2fb24bedc6b4555f41da611ff4d0e4/sandbox/linux-sandbox/1030/execroot/nu_workforce/bazel-out/k8-opt-exec-2B5CBBC6-ST-625e526ca8a8/bin/external/aspect_rules_js/js/private/js_image_layer_builder.sh.runfiles/aspect_rules_js/js/private/js_image_layer.mjs:9719:30)
Target //mytarget:image failed to build
```
I've added some trace in the build function around [line 9712](https://github.com/aspect-build/rules_js/blob/f284cb6f0434a3114fce30c70429a15646c409b6/js/private/js_image_layer.mjs#L9712) :
```
root: bazel-out/k8-fastbuild-ST-4a519fd6d3e4/bin
dest: bazel-out/k8-fastbuild-ST-4a519fd6d3e4/bin/external/com_google_protobuf/_virtual_imports/any_proto/google/protobuf/any.proto
realp: /home/jpgariep/.cache/bazel/_bazel_jpgariep/dc2fb24bedc6b4555f41da611ff4d0e4/external/com_google_protobuf/src/google/protobuf/any.proto
output_path: o
```
It seems that the code is assuming that `root` will be somewhere in `realp` but it's not the case here.

I'm building images according to the [documentation here - w/ legacy rules_docker](https://docs.aspect.build/rules/aspect_rules_js/docs/js_image_layer/)

### Version

Development (host) and target OS/architectures:
linux/x86_64

Output of `bazel --version`:
bazel 6.0.0

Version of the Aspect rules, or other relevant rules from your
`WORKSPACE` or `MODULE.bazel` file:

v1.20.1

Language(s) and/or frameworks involved:

TypeScript

### How to reproduce

https://github.com/gawi/rules_js/tree/main/e2e/js_image_docker

### Any other information?

_No response_

### Fund our work

- [ ] Sponsor our open source work by donating a [bug bounty](https://opencollective.com/aspect-build/)

### PR DESCRIPTION
**Pull request by thesayyn**
js_image_layer treated external repository symlinks to the sandbox as if it was a symlink created by `ctx.actions.symlink` which failed as they did not have a corresponding path in the manifest. 

### Type of change

- Bug fix (change which fixes an issue)  fixes #920, #1253, #1260

### Test plan

- New test cases added

### CODE REVIEW

File: .bazelversion [hunk 1]
```diff
@@ -1,4 +1,4 @@
-6.1.2
+5.4.1
```
Comment (jbedard) [inline .bazelversion:1]: Revert before merge?

File: e2e/js_image_oci/WORKSPACE [hunk 2]
```diff
@@ -87,3 +87,38 @@ http_archive(
 load("@container_structure_test//:repositories.bzl", "container_structure_test_register_toolchain")
 
 container_structure_test_register_toolchain(name = "container_structure_test_toolchain")
+
+## Testing
```
Comment (jbedard) [inline e2e/js_image_oci/WORKSPACE:91]: Isn't this entire WORKSPACE for testing?

File: e2e/js_image_oci/repo/BUILD.bazel [hunk 3]
```diff
@@ -0,0 +1,29 @@
+load("@aspect_bazel_lib//lib:copy_to_directory.bzl", "copy_to_directory")
+load("@aspect_bazel_lib//lib:copy_to_bin.bzl", "copy_to_bin")
+
+package(default_visibility = ["//visibility:public"])
+
+filegroup(
+    name = "sources",
+    srcs = [
+        "source.txt",
+        "srcdir/source.txt",
```
Comment (jbedard) [inline e2e/js_image_oci/repo/BUILD.bazel:10]: Any reason to name them the same?

File: js/private/image/index.ts [hunk 4]
```diff
@@ -28,15 +29,46 @@ type Entries = { [path: string]: Entry }
...
             return key
         }
     }
-    throw new Error(
-        `couldn't map ${value} to a path. please file a bug at https://github.com/aspect-build/rules_js/issues/new/choose`
-    )
+    return undefined
+}
+
+async function isSymbolicLink(p: string) {
```
Comment (jbedard) [inline js/private/image/index.ts:41]: This returns more then a boolean so shouldn't have the `is*` naming. Can this just be `readlink` or `readBazelLink` or something like that?

File: js/private/image/index.ts [hunk 5]
```diff
@@ -28,15 +29,46 @@ type Entries = { [path: string]: Entry }
...
+                return firstHop
+            }
+            try {
+                const secondHop = await readlink(firstHop)
+                // if the first hop is not equal to second hop then it means
+                // this is symlink generated by ctx.actions.symlink
+                if (secondHop != firstHop) {
+                    return secondHop
+                }
+            } catch (e) {
```
Comment (jbedard) [inline js/private/image/index.ts:59]: The outer catch does the same thing and I think we can just drop this `catch` block?

File: e2e/js_image_oci/WORKSPACE (hunk 2 above)
Comment (thesayyn) [reply e2e/js_image_oci/WORKSPACE:91]: Yes, it is 👯  kind of pointless comment. will revert

File: js/private/image/index.ts [hunk 6]
```diff
@@ -28,15 +29,54 @@ type Entries = { [path: string]: Entry }
...
             return key
         }
     }
-    throw new Error(
-        `couldn't map ${value} to a path. please file a bug at https://github.com/aspect-build/rules_js/issues/new/choose`
-    )
+    return undefined
+}
+
+async function isSymbolicLink(filep: string, root: string) {
```
Comment (jbedard) [inline js/private/image/index.ts:41]: This seems to return a string and not a boolean like I'd expect an `is*` method to. Can we rename this?

File: js/private/image/index.ts (hunk 5 above)
Comment (thesayyn) [reply js/private/image/index.ts:67]: right...

File: js/private/image/index.ts (hunk 6 above)
Comment (thesayyn) [reply js/private/image/index.ts:41]: agreed. renaming to get

File: e2e/js_image_oci/repo/BUILD.bazel (hunk 3 above)
Comment (thesayyn) [reply e2e/js_image_oci/repo/BUILD.bazel:10]: no, just wanted to assert that external directories are treated okay.

File: js/private/image/index.ts [hunk 7]
```diff
@@ -237,15 +267,28 @@ export async function build(
...
         }
 
-        const realp = await realpath(dest)
-        const output_path = realp.slice(realp.indexOf(root))
-        if (output_path != dest) {
+        const realp = await getSymbolicLink(dest, root)
+
+        // it's important that we don't treat any symlink pointing out of execroot since
+        // bazel symlinks external files into sandbox to make them available to us.
+        if (realp && !is_external) {
```
Comment (jbedard) [inline js/private/image/index.ts:274]: nit: avoid `getSymbolicLink` if `is_external`?

File: js/private/image/index.ts [hunk 8]
```diff
@@ -28,15 +29,43 @@ type Entries = { [path: string]: Entry }
...
+
+
+
+async function getSymbolicLink(filep: string, root: string) {
+    const p = path.resolve(filep)
+    try {
+        let prevHop = p;
+        let hopped = 0;
+        while (true) {
+            let nextHop = realpathSync(prevHop);
```
Comment (jbedard) [inline js/private/image/index.ts:49]: nit: use `async`?

Comment (jbedard) [inline js/private/image/index.ts:49]: double check `realpath` always returns a resolved (non-relative) path

File: js/private/image/index.ts [hunk 9]
```diff
@@ -28,15 +29,43 @@ type Entries = { [path: string]: Entry }
...
+                prevHop = nextHop;
+                hopped++;
+            } else if(hopped == 0) {
+                return undefined;
+            } else {
+                return nextHop;
+            }
+        }
+    } catch (e) {
+        if (e.code != 'EINVAL') {
```
Comment (jbedard) [inline js/private/image/index.ts:63]: Needed? Drop the try/catch?

**Review by jbedard — APPROVED** [review]

File: js/private/image/index.ts [hunk 10]
```diff
@@ -28,15 +29,83 @@ type Entries = { [path: string]: Entry }
...
+        }
+    }
+}
+
+async function resolveSymlink(p: string) {
+    let prevHop = path.resolve(p);
+    let hopped = false;
+ 
+    while (true) {
+        // /output-base/sandbox/4/execroot/wksp/bazel-out
```
Comment (jbedard) [inline js/private/image/index.ts:94]: Delete these 2 comments?

File: js/private/image/index.ts [hunk 11]
```diff
@@ -116,7 +185,7 @@ function add_symlink(
     pkg: Pack,
     stats: HermeticStat
 ) {
-    const link_parent = path.dirname(name);
+    const link_parent = path.dirname(name)
```
Comment (jbedard) [inline js/private/image/index.ts:188]: 😲

**Review by jbedard — APPROVED** [review]

**Pull request merged**